web_src/js/features/user-auth-webauthn.js

```diff
@@ -0,0 +1,240 @@
+const {appSubUrl, csrfToken} = window.config;
+
+
+export function initUserAuthWebAuthn() {
+  if ($('.user.signin.webauthn-prompt').length === 0) {
+    return;
+  }
+
+  if (!detectWebAuthnSupport()) {
+    return;
+  }
+
+  $.getJSON(`${appSubUrl}/user/webauthn/assertion`, {})
+    .done((makeAssertionOptions) => {
+      makeAssertionOptions.publicKey.challenge = bufferDecode(makeAssertionOptions.publicKey.challenge);
+      for (let i = 0; i < makeAssertionOptions.publicKey.allowCredentials.length; i++) {
+        makeAssertionOptions.publicKey.allowCredentials[i].id = bufferDecode(makeAssertionOptions.publicKey.allowCredentials[i].id);
+      }
+      navigator.credentials.get({
+        publicKey: makeAssertionOptions.publicKey
+      })
+        .then((credential) => {
+          verifyAssertion(credential);
+        }).catch((err) => {
+          webAuthnError(0, err.message);
+        });
+    }).fail(() => {
+      webAuthnError('unknown');
+    });
+}
+
+function verifyAssertion(assertedCredential) {
+  // Move data into Arrays incase it is super long
+  const authData = new Uint8Array(assertedCredential.response.authenticatorData);
+  const clientDataJSON = new Uint8Array(assertedCredential.response.clientDataJSON);
+  const rawId = new Uint8Array(assertedCredential.rawId);
+  const sig = new Uint8Array(assertedCredential.response.signature);
+  const userHandle = new Uint8Array(assertedCredential.response.userHandle);
+  $.ajax({
+    url: `${appSubUrl}/user/webauthn/assertion`,
+    type: 'POST',
+    data: JSON.stringify({
+      id: assertedCredential.id,
+      rawId: bufferEncode(rawId),
+      type: assertedCredential.type,
+      response: {
+        authenticatorData: bufferEncode(authData),
+        clientDataJSON: bufferEncode(clientDataJSON),
+        signature: bufferEncode(sig),
+        userHandle: bufferEncode(userHandle),
+      },
+    }),
+    contentType: 'application/json; charset=utf-8',
+    dataType: 'json',
+    success: (resp) => {
+      if (resp && resp['redirect']) {
+        window.location.href = resp['redirect'];
+      } else {
+        window.location.href = '/';
+      }
+    },
+    error: (xhr) => {
+      if (xhr.status === 500) {
+        webAuthnError('unknown');
+        return;
+      }
+      webAuthnError('unable-to-process');
+    }
+  });
+}
+
+const lookup = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
+
+function encode (num) {
+  return lookup.charAt(num);
+}
+
+function tripletToBase64 (num) {
```

Comment:
can this func get a test?

Reply:
Obsolete comment, function is gone with the move to the dependency.